### PR TITLE
Implementation of nullable datatypes

### PR DESCRIPTION
**Pull request by fwilhelm92**
**Description**

Implementation of nullable datatypes

**Tests**

- [x] make lint
- [x] make test

### CODE REVIEW

File: tds/field.go [hunk 1]
```diff
@@ -86,6 +86,9 @@ type FieldFmt interface {
 	// Format information as sent to or received from TDS server
 	DataType() asetypes.DataType
 	setDataType(asetypes.DataType)
+	// ConvertDataType converts a datatype to the passed nullable
+	// datatype and sets according status and length
+	ConvertDataType(asetypes.DataType)
```
Comment (ntnn) [inline tds/field.go:91]: Does the passed datatype have to be nullable? Looking at the code at a glance it doesn't seem like it.

File: term/helpers.go [hunk 2]
```diff
@@ -148,13 +148,18 @@ func processRows(rows driver.Rows) error {
...
+				switch rowsColumnTypeName.ColumnTypeDatabaseTypeName(i) {
+				case "DECIMAL", "DECN", "NUMN":
+					cellS = cell.(*asetypes.Decimal).String()
+				case "IMAGE", "BINARY", "LONGBINARY", "VARBINARY":
+					cellS = hex.EncodeToString(cell.([]byte))
+				default:
+					cellS = fmt.Sprintf("%v", (interface{})(cell))
+				}
+			} else {
+				cellS = "<nil>"
```
Comment (ntnn) [inline term/helpers.go:162]:
```suggestion
			if cell == nil {
				cellS = "<nil>"
			} else {
				switch rowsColumnTypeName.ColumnTypeDatabaseTypeName(i) {
				case "DECIMAL", "DECN", "NUMN":
					cellS = cell.(*asetypes.Decimal).String()
				case "IMAGE", "BINARY", "LONGBINARY", "VARBINARY":
					cellS = hex.EncodeToString(cell.([]byte))
				default:
					cellS = fmt.Sprintf("%v", (interface{})(cell))
				}
```

The simpler branch should come first for readability.

File: asetypes/nullableTypes.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+	TIME:       TIMEN,
+	UINT2:      UINTN,
+	UINT4:      UINTN,
+	UINT8:      UINTN,
+}
+
+func (t DataType) NullableType() (DataType, error) {
+	nullableType, ok := TypeToNullableTypes[t]
+	if !ok {
+		return t, fmt.Errorf("nullable datatype for %s not supported", t)
```
Comment (ntnn) [inline asetypes/nullableTypes.go:129]:
```suggestion
		return t, fmt.Errorf("datatype %q has no nullable equivalent", t)
```

The current message sounds like the implementation doesn't support the datatype as nullable.

File: asetypes/nullableTypes.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+// https://cs.opensource.google/go/go/+/refs/tags/go1.17.2:src/database/sql/convert.go;l=210
+func (nb *NullBinary) Scan(value interface{}) error {
+	// Replaces sql.convertAssign-functionality of checking
+	// nil-destination
+	if nb == nil {
+		return fmt.Errorf("destination pointer is nil")
+	}
+
+	if value == nil {
+		nb.ByteSlice, nb.Valid = []byte{}, false
```
Comment (ntnn) [inline asetypes/nullableTypes.go:84]:
```suggestion
		nb.ByteSlice = []byte{}
		nb.Valid = false
```

Multiple assignments may work like that but they're going against KISS.

File: asetypes/nullableTypes.go [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+// https://cs.opensource.google/go/go/+/refs/tags/go1.17.2:src/database/sql/convert.go;l=210
+func (nd *NullDecimal) Scan(value interface{}) error {
+	// Replaces sql.convertAssign-functionality of checking
+	// nil-destination
+	if nd == nil {
+		return fmt.Errorf("destination pointer is nil")
+	}
+
+	if value == nil {
+		nd.Dec, nd.Valid = &Decimal{}, false
```
Comment (ntnn) [inline asetypes/nullableTypes.go:37]:
```suggestion
		nd.Dec = &Decimal{}
		nd.Valid = false
```

KISS

**Review by ntnn — CHANGES_REQUESTED** [review]

File: tds/field.go (hunk 1 above)
Comment (fwilhelm92) [reply tds/field.go:91]: Change to `ConvertToNullableDataType`

**Review by cb80 — APPROVED** [review]
Impressive!

File: asetypes/convert.go [hunk 6]
```diff
@@ -7,75 +7,62 @@
...
-			return sv.Int(), nil
-		}
-	case UINT2:
-		switch sv.Kind() {
-		case reflect.Int, reflect.Int64, reflect.Int32, reflect.Int16, reflect.Int8,
-			reflect.Uint64, reflect.Uint32, reflect.Uint16, reflect.Uint8:
-			return uint16(sv.Uint()), nil
+	// Check for driver.Valuer interface, e.g. database/sql.Null<types>
+	vv, ok := v.(driver.Valuer)
+	if ok {
```
Comment (ntnn) [inline asetypes/convert.go:32]:
```suggestion
	if vv, ok := v.(driver.Valuer); ok {
```

**Review by ntnn — CHANGES_REQUESTED** [review]

**Review by ntnn — APPROVED** [review]

**Pull request merged**